src/main/java/com/aws/greengrass/shadowmanager/sync/strategy/BaseSyncStrategy.java

```diff
@@ -198,10 +219,51 @@ public void stop() {
         }
     }
 
-
     abstract void doStart(SyncContext context, int syncParallelism);
 
-    abstract void doStop();
+    protected void doStop() {
+        logger.atInfo(SYNC_EVENT_TYPE).log("Cancel {} sync thread(s)", syncThreads.size());
+        try {
+            // The syncing flag is set to false by this point so once current requests have finished threads will not
+            // pick up another
+
+            // We need to cancel threads, but we should wait for any inflight requests to finish
+            // Cancelling a thread while the db is in the middle of an operation will cause errors in the connection.
+            // FullSyncRequests can generate a local and cloud request and we want both of those to get executed before
+            // cancelling the sync threads.
+
+            // There is a semaphore around the critical section in the sync loop - we do not want to cancel while
+            // that is held.
+            // There is a latch that is decremented when the syncLoop exits - this allows us to know when stopped or
+            // cancelled threads have actually exited. Cancelling a future will not let you know when the thread has
+            // actually finished.
+            if (!exec.tryAcquire(syncParallelism)) {
+                logger.atInfo(SYNC_EVENT_TYPE).log("Waiting for in-flight sync requests to finish");
+                exec.acquireUninterruptibly(syncParallelism);
```

Comment:
Why not just call this instead of `tryAcquire` and then `acquireUninterruptibly`?

Reply:
I wanted to log that it's going to wait - if I just called `acquireUniterruptibly` there's no opportunity to print a message